bigint/src/lib.rs

```diff
@@ -1349,6 +1351,40 @@ impl Integer for BigUint {
     }
 }
 
+fn high_bits_to_u64(v: &BigUint) -> u64 {
+    match v.data.len() {
+        0   => 0,
+        1   => v.data[0] as u64,
+        _   => {
+            let mut bits = v.bits();
+            let mut ret = 0u64;
+            let mut ret_bits = 0;
+
+            for d in v.data.iter().rev() {
+                let digit_bits = ((bits - 1) & (big_digit::BITS - 1)) + 1;
```

Comment:
AKA `bits % BITS`, right?  Just write that, and let the optimizer be the clever one.